File: ts/src/values/builtinFunctionValue.ts
import VirtualMachine from "../virtualMachine";
import ArrayValue from "./arrayValue";
import { CompareResult, IFunctionValue, IValue } from "./ivalues";

export type BuiltinFunctionCallback = (vm: VirtualMachine, args: ArrayValue) => void;

export default class BuiltinFunctionValue implements IFunctionValue
{
    public readonly value: BuiltinFunctionCallback;
    public readonly name: string;

    constructor (value: BuiltinFunctionCallback, name: string)
    {
        this.value = value;
        this.name = name;
    }

    public typename() { return 'builtin-function'; }
    public toString() { return `builtin-function:${this.name}`; }

    public compareTo(other: IValue): CompareResult
    {
        if (!(other instanceof BuiltinFunctionValue)) { return 1; }

        return this.value === other.value ? 0 : 1;
    }

    public invoke(vm: VirtualMachine, args: ArrayValue, pushToStackTrace: boolean)
    {
        this.value(vm, args);
    }
}

File: unity/simple-stack-vm-unity/Assets/Scripts/SimpleStackVM/Values/BuiltinFunctionValue.cs
#nullable enable

using System.Runtime.CompilerServices;

namespace SimpleStackVM
{
    public struct BuiltinFunctionValue : IFunctionValue
    {
        #region Field
        public delegate void BuiltinFunctionDelegate(VirtualMachine vm, ArrayValue args);

        public readonly BuiltinFunctionDelegate Value;

        public string TypeName => "builtin-function";
        #endregion

        #region Constructor
        public BuiltinFunctionValue(BuiltinFunctionDelegate value)
        {
            this.Value = value;
        }
        #endregion

        #region Methods
        public int CompareTo(IValue? other)
        {
            if (other == null || !(other is BuiltinFunctionValue otherFunction))
            {
                return 1;
            }

            return this.Value == otherFunction.Value ? 0 : 1;
        }

        public override string ToString() => "builtin-function";

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public void Invoke(VirtualMachine vm, ArrayValue args, bool pushToStackTrace)
        {
            this.Value.Invoke(vm, args);
        }
        #endregion
    }
}